### PR TITLE
Model: Run user callbacks before history and logger

### PR DESCRIPTION
**Pull request by pasky**
This allows user callbacks to generate and store extra accuracy statistics on the validation set, like MRR/MAP for ranking tasks.

I'm not sure if this is the best solution, even though it doesn't seem to break anything here.

### CODE REVIEW

**Comment by fchollet** [comment]
This is in conflict with https://github.com/fchollet/keras/pull/1683/files

Thoughts on which we should merge (if any)?

**Comment by gw0** [comment]
Adding custom metrics/statistics with callbacks is needed, but having an average `loss` metric available at `on_epoch_end()` (not just `val_loss`) is also useful. I haven't stumbled upon this, because I was using an alternative history-like callback `CSVHistory` (closed PR #1684).

Solutions for both:

- `History` callback should be split into two parts, one for computing average metrics `on_epoch_end()` (before user callbacks) and the other for keeping track of history and totals metrics `on_batch_*()` (after user callbacks)
- `models.py` should handle computing average metrics generated per batch by Keras (`loss`, `acc`), this means moving the totals computation out of `History` callback and into `models.py`

**Comment by pasky** [comment]
I'm not actually using the History for anything but my custom metrics, but it makes sense to me that it's probably not necessary to enforce the current structure of the History callback and I like the outlined solution.

So, perhaps the smoothest option would be to merge this and update #1683 to rather move the averages computation outside of the callbacks.  But I don't want to forcefully preempt this :) - @gw0, would that make sense to you?

**Comment by gw0** [comment]
@pasky, sure. @fchollet which variant do you prefer - split `History` callback into two parts or average computation in `models.py`?

**Comment by fchollet** [comment]
This is fixed by moving metrics averages computation to `BaseLogger`, and moving `History` and `ProgbarLogger` to the end of the callbacks list.
